cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -2076,4 +2099,20 @@ private String getLatestAppVersion(NamespaceId namespaceId, String appId) throws
     }
     return latestApplicationMeta.getSpec().getAppVersion();
   }
+
+  /**
+   * Convert BatchProgram to ProgramId with the latest App version.
+   * Default version will be used if application can not be found.
+   */
+  private ProgramId batchProgramToProgramId(String namespace, BatchProgram batchProgram) {
+    try {
+      ApplicationId applicationId = new ApplicationId(namespace, batchProgram.getAppId(),
+                                                      getLatestAppVersion(new NamespaceId(namespace),
+                                                                          batchProgram.getAppId()));
+      return new ProgramId(applicationId, batchProgram.getProgramType(), batchProgram.getProgramId());
+    } catch (ApplicationNotFoundException e) {
+      return new ProgramId(namespace, batchProgram.getAppId(), batchProgram.getProgramType(),
```

Comment:
Why not just throw `ApplicationNotFoundException` when the app does not exist?

Reply:
Also it feels inconsistent to remain putting back default `-SNAPSHOT` version here, while we're using latest version in other places in this PR

Reply:
agree with above

Reply:
Agree it is bit confusing, but we do need a default ProgramId as a place holder because ProgramId is part of the BatchProgramResult even if we can't find the program - https://github.com/cdapio/cdap/blob/e8e67361dbf78c5c967800c87370d2dc8b5a74ec/cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java#L1330-L1332

I have added some comments to make this clear, but let me know if you have a better work around for this.